overlays/turnkey.d/sslcert/usr/local/bin/turnkey-make-ssl-cert

```diff
@@ -270,25 +299,57 @@ else
   CSR="${CERT%.*}.csr"
 fi
 
-# don't overwrite existing file without permission
-if [[ -f $CERT ]] && [[ $OVERWRITE == false ]]; then
-  error_exit "Output file already exists: $CERT"
+
+
+# don't overwrite existing key and cert files without permission
+if [[ -f $CERT ]] && [[ -f $KEY ]]; then
+    # except when only generating dhparams
+    if [[ $OVERWRITE == false ]] && [[ $DH_ONLY == false ]]; then
+        error_exit "Output file already exists: $CERT"
+    fi
+elif [[ $DH_ONLY == true ]]; then
+    msg="$CERT & $KEY files don't exist, won't be able to rebuild combined"
+    error_exit "$msg cert.pem file. Please rerun with different options."
 fi
 
 # create the key and certificate.
-if ! openssl req -sha256 -config $TMPFILE -new -x509 -days $days -nodes \
-  -out $CERT -keyout $KEY > $TMPOUT 2>&1
-then
-  error_exit "Could not create certificate. Openssl output was:" $TMPOUT;
+if [[ $DH_ONLY == false ]]; then
+    info "Generating certificate and key files."
+    if ! openssl req -sha256 -config $TMPFILE -new -x509 -days $days \
+            -nodes -out $CERT -keyout $KEY > $TMPOUT 2>&1; then
+        error_exit "Could not create certificate. Openssl output was:" $TMPOUT;
+    fi
+fi
+
+# gen dhparams.
+if [[ "$GEN_DH" == "true" ]]; then
+    if [[ $DH_BITS -le 1050 ]]; then
+        _time="a while"
+    elif [[ $DH_BITS -le 3000 ]]; then
+        _time="a long time"
+    else
+        _time="hours, possibly days"
+    fi
+    info "Generating Diffie-Hellman parameters file - this may take $_time..."
+    if ! openssl dhparam -out $DHP $DH_BITS > $TMPOUT 2>&1; then
+        msg="Diffie-Hellman parameter generation failed."
+        error_exit "$msg OpenSSL output: " $TMPOUT
+    elif [[ $DH_BITS -lt 2048 ]]; then
+        msg="Diffie-Hellman parameter bit size is $DH_BITS. Recommended"
+        msg="$msg minimum is $DH_BITS_REC; 4096 is best, but may take hours to"
+        warning "$msg generate."
```

Comment:
This seems a little odd to me the way you're creating this I would've thought it cleaner to simply do

```bash
msg="Diffie-Hellman parameter bit size is $DH_BITS. Recommended
minimum is $DH_BITS_REC; 4096 is best, but may take hours to
generate"
```

(fwiw this will replace newlines with spaces, but in this case that appears to be intended)

Reply:
Well spotted. That is certainly much nicer looking. FWIW, there are a couple more places that could be tidied, so I did them too! :smile: